File: sdks/python/src/cvfile/_xmp.py
"""XMP metadata writer + reader for the cv: namespace."""

from __future__ import annotations

import json
import re
from dataclasses import asdict
from datetime import datetime, timezone
from typing import Any

from cvfile._constants import CV_NAMESPACE_URI, CV_SPEC_VERSION, DEFAULT_GENERATOR
from cvfile._types import AlternateMeta, CvMetadata, EmbeddingSpaceSummary, IntegrityEntry

XMP_BEGIN = '<?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>'
XMP_END = '<?xpacket end="w"?>'


def _iso(dt: datetime) -> str:
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    s = dt.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    return s


def _xml_escape(s: str) -> str:
    return (
        s.replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
        .replace("'", "&apos;")
    )


def _xml_unescape(s: str) -> str:
    return (
        s.replace("&apos;", "'")
        .replace("&quot;", '"')
        .replace("&gt;", ">")
        .replace("&lt;", "<")
        .replace("&amp;", "&")
    )


def build_xmp(
    *,
    version: str,
    primary_language: str,
    primary_payload: str,
    created: datetime,
    modified: datetime | None = None,
    generator: str = DEFAULT_GENERATOR,
    alternates: tuple[AlternateMeta, ...] = (),
    integrity: tuple[IntegrityEntry, ...] = (),
    embeddings: tuple[EmbeddingSpaceSummary, ...] = (),
) -> str:
    """Build an XMP packet with the cv: namespace and PDF/A-3u identification."""
    modified = modified or created

    def _alt_dict(a: AlternateMeta) -> dict[str, str]:
        return {"payload": a.payload, "language": a.language, "mimeType": a.mime_type}

    def _int_dict(i: IntegrityEntry) -> dict[str, str]:
        return {"payload": i.payload, "algorithm": i.algorithm, "digest": i.digest}

    def _emb_dict(e: EmbeddingSpaceSummary) -> dict[str, Any]:
        return {"model": e.model, "dimension": e.dimension, "metric": e.metric, "chunks": e.chunks}

    alt_block = (
        f'\n      <cv:alternates>{_xml_escape(json.dumps([_alt_dict(a) for a in alternates]))}</cv:alternates>'
        if alternates
        else ""
    )
    int_block = (
        f'\n      <cv:integrity>{_xml_escape(json.dumps([_int_dict(i) for i in integrity]))}</cv:integrity>'
        if integrity
        else ""
    )
    emb_block = (
        f'\n      <cv:embeddings>{_xml_escape(json.dumps([_emb_dict(e) for e in embeddings]))}</cv:embeddings>'
        if embeddings
        else ""
    )

    ext_schema = _cv_extension_schema()

    return f"""{XMP_BEGIN}
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="cvfile-py {_xml_escape(CV_SPEC_VERSION)}">
  <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <rdf:Description rdf:about=""
      xmlns:pdfaid="http://www.aiim.org/pdfa/ns/id/">
      <pdfaid:part>3</pdfaid:part>
      <pdfaid:conformance>U</pdfaid:conformance>
    </rdf:Description>
    <rdf:Description rdf:about=""
      xmlns:dc="http://purl.org/dc/elements/1.1/">
      <dc:format>application/pdf</dc:format>
    </rdf:Description>
    <rdf:Description rdf:about=""
      xmlns:xmp="http://ns.adobe.com/xap/1.0/">
      <xmp:CreateDate>{_xml_escape(_iso(created))}</xmp:CreateDate>
      <xmp:ModifyDate>{_xml_escape(_iso(modified))}</xmp:ModifyDate>
      <xmp:CreatorTool>{_xml_escape(generator)}</xmp:CreatorTool>
    </rdf:Description>
    <rdf:Description rdf:about=""
      xmlns:cv="{CV_NAMESPACE_URI}">
      <cv:version>{_xml_escape(version)}</cv:version>
      <cv:created>{_xml_escape(_iso(created))}</cv:created>
      <cv:modified>{_xml_escape(_iso(modified))}</cv:modified>
      <cv:primaryLanguage>{_xml_escape(primary_language)}</cv:primaryLanguage>
      <cv:primaryPayload>{_xml_escape(primary_payload)}</cv:primaryPayload>
      <cv:generator>{_xml_escape(generator)}</cv:generator>{alt_block}{int_block}{emb_block}
    </rdf:Description>
{ext_schema}
  </rdf:RDF>
</x:xmpmeta>
{XMP_END}"""


def _cv_extension_schema() -> str:
    props = [
        ("version", "Text", "cvfile.org format version (MAJOR.MINOR)"),
        ("created", "Date", "When the .cv file was created"),
        ("modified", "Date", "When the .cv file was last modified"),
        ("primaryLanguage", "Text", "BCP-47 tag of the canonical content language"),
        ("primaryPayload", "Text", "Filename of the canonical text payload"),
        ("generator", "Text", "Identifier of the producer"),
        ("alternates", "Text", "Alternate payload descriptors (JSON-encoded array)"),
        ("integrity", "Text", "Per-payload digest entries (JSON-encoded array)"),
        ("embeddings", "Text", "Embedding-space summaries (JSON-encoded array)"),
    ]

    def _prop_xml(name: str, value_type: str, desc: str) -> str:
        return f"""                <rdf:li rdf:parseType="Resource">
                  <pdfaProperty:name>{name}</pdfaProperty:name>
                  <pdfaProperty:valueType>{value_type}</pdfaProperty:valueType>
                  <pdfaProperty:category>external</pdfaProperty:category>
                  <pdfaProperty:description>{_xml_escape(desc)}</pdfaProperty:description>
                </rdf:li>"""

    items = "\n".join(_prop_xml(*p) for p in props)
    return f"""    <rdf:Description rdf:about=""
      xmlns:pdfaExtension="http://www.aiim.org/pdfa/ns/extension/"
      xmlns:pdfaSchema="http://www.aiim.org/pdfa/ns/schema#"
      xmlns:pdfaProperty="http://www.aiim.org/pdfa/ns/property#">
      <pdfaExtension:schemas>
        <rdf:Bag>
          <rdf:li rdf:parseType="Resource">
            <pdfaSchema:namespaceURI>{CV_NAMESPACE_URI}</pdfaSchema:namespaceURI>
            <pdfaSchema:prefix>cv</pdfaSchema:prefix>
            <pdfaSchema:schema>cvfile.org cv namespace</pdfaSchema:schema>
            <pdfaSchema:property>
              <rdf:Seq>
{items}
              </rdf:Seq>
            </pdfaSchema:property>
          </rdf:li>
        </rdf:Bag>
      </pdfaExtension:schemas>
    </rdf:Description>"""


_TAG_RE_CACHE: dict[str, re.Pattern[str]] = {}


def _tag_re(tag: str) -> re.Pattern[str]:
    cached = _TAG_RE_CACHE.get(tag)
    if cached is None:
        cached = re.compile(rf"<{re.escape(tag)}>([\s\S]*?)</{re.escape(tag)}>")
        _TAG_RE_CACHE[tag] = cached
    return cached


def _inner(xml: str, tag: str) -> str | None:
    m = _tag_re(tag).search(xml)
    return _xml_unescape(m.group(1).strip()) if m else None


def _parse_json_text(xml: str, tag: str) -> Any:
    raw = _inner(xml, tag)
    if not raw:
        return None
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return None


def parse_xmp(xml: str) -> CvMetadata | None:
    """Parse a cv:* XMP packet. Returns None if required fields are missing."""
    version = _inner(xml, "cv:version")
    primary_language = _inner(xml, "cv:primaryLanguage")
    primary_payload = _inner(xml, "cv:primaryPayload")
    if not version or not primary_language or not primary_payload:
        return None

    created_str = _inner(xml, "cv:created")
    modified_str = _inner(xml, "cv:modified")
    generator = _inner(xml, "cv:generator")

    def _parse_dt(s: str | None) -> datetime | None:
        if not s:
            return None
        try:
            return datetime.fromisoformat(s.replace("Z", "+00:00"))
        except ValueError:
            return None

    raw_alts = _parse_json_text(xml, "cv:alternates") or []
    alternates = tuple(
        AlternateMeta(payload=a["payload"], language=a["language"], mime_type=a["mimeType"])
        for a in raw_alts
    )

    raw_ints = _parse_json_text(xml, "cv:integrity") or []
    integrity = tuple(
        IntegrityEntry(payload=i["payload"], algorithm=i["algorithm"], digest=i["digest"])
        for i in raw_ints
    )

    raw_embs = _parse_json_text(xml, "cv:embeddings") or []
    embeddings = tuple(
        EmbeddingSpaceSummary(model=e["model"], dimension=e["dimension"], metric=e["metric"], chunks=e["chunks"])
        for e in raw_embs
    )

    return CvMetadata(
        version=version,
        primary_language=primary_language,
        primary_payload=primary_payload,
        created=_parse_dt(created_str),
        modified=_parse_dt(modified_str),
        generator=generator,
        alternates=alternates,
        integrity=integrity,
        embeddings=embeddings,
    )


__all__ = ["build_xmp", "parse_xmp"]

File: packages/sdk-js/src/xmp.ts
import { CV_NAMESPACE_URI, CV_SPEC_VERSION, DEFAULT_GENERATOR } from './constants.js';
import type { AlternateMeta, CvMetadata, EmbeddingSpaceSummary, IntegrityEntry } from './types.js';

const XMP_BEGIN = '<?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>';
const XMP_END = '<?xpacket end="w"?>';

export interface XmpInput {
  version: string;
  primaryLanguage: string;
  primaryPayload: string;
  created: Date;
  modified?: Date;
  generator?: string;
  alternates?: AlternateMeta[];
  integrity?: IntegrityEntry[];
  embeddings?: EmbeddingSpaceSummary[];
}

export function buildXmp(input: XmpInput): string {
  const escape = xmlEscape;
  const generator = input.generator ?? DEFAULT_GENERATOR;
  const created = isoDate(input.created);
  const modified = input.modified ? isoDate(input.modified) : created;

  // Structured collections are serialized as JSON-encoded Text values inside the
  // XMP packet so that the cv: extension schema can declare them as simple Text
  // (PDF/A-3u rule 6.6.2.3.1 forbids unnamed structured types). Consumers
  // recover the structure via JSON.parse.
  const altBlock = (input.alternates ?? []).length > 0
    ? `\n      <cv:alternates>${escape(JSON.stringify(input.alternates))}</cv:alternates>`
    : '';
  const intBlock = (input.integrity ?? []).length > 0
    ? `\n      <cv:integrity>${escape(JSON.stringify(input.integrity))}</cv:integrity>`
    : '';
  const embBlock = (input.embeddings ?? []).length > 0
    ? `\n      <cv:embeddings>${escape(JSON.stringify(input.embeddings))}</cv:embeddings>`
    : '';

  return `${XMP_BEGIN}
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="cvfile-sdk ${escape(CV_SPEC_VERSION)}">
  <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <rdf:Description rdf:about=""
      xmlns:pdfaid="http://www.aiim.org/pdfa/ns/id/">
      <pdfaid:part>3</pdfaid:part>
      <pdfaid:conformance>U</pdfaid:conformance>
    </rdf:Description>
    <rdf:Description rdf:about=""
      xmlns:dc="http://purl.org/dc/elements/1.1/">
      <dc:format>application/pdf</dc:format>
    </rdf:Description>
    <rdf:Description rdf:about=""
      xmlns:xmp="http://ns.adobe.com/xap/1.0/">
      <xmp:CreateDate>${escape(created)}</xmp:CreateDate>
      <xmp:ModifyDate>${escape(modified)}</xmp:ModifyDate>
      <xmp:CreatorTool>${escape(generator)}</xmp:CreatorTool>
    </rdf:Description>
    <rdf:Description rdf:about=""
      xmlns:cv="${CV_NAMESPACE_URI}">
      <cv:version>${escape(input.version)}</cv:version>
      <cv:created>${escape(created)}</cv:created>
      <cv:modified>${escape(modified)}</cv:modified>
      <cv:primaryLanguage>${escape(input.primaryLanguage)}</cv:primaryLanguage>
      <cv:primaryPayload>${escape(input.primaryPayload)}</cv:primaryPayload>
      <cv:generator>${escape(generator)}</cv:generator>${altBlock}${intBlock}${embBlock}
    </rdf:Description>
${cvExtensionSchema()}
  </rdf:RDF>
</x:xmpmeta>
${XMP_END}`;
}

function cvExtensionSchema(): string {
  return `    <rdf:Description rdf:about=""
      xmlns:pdfaExtension="http://www.aiim.org/pdfa/ns/extension/"
      xmlns:pdfaSchema="http://www.aiim.org/pdfa/ns/schema#"
      xmlns:pdfaProperty="http://www.aiim.org/pdfa/ns/property#">
      <pdfaExtension:schemas>
        <rdf:Bag>
          <rdf:li rdf:parseType="Resource">
            <pdfaSchema:namespaceURI>${CV_NAMESPACE_URI}</pdfaSchema:namespaceURI>
            <pdfaSchema:prefix>cv</pdfaSchema:prefix>
            <pdfaSchema:schema>cvfile.org cv namespace</pdfaSchema:schema>
            <pdfaSchema:property>
              <rdf:Seq>
${CV_EXTENSION_PROPERTIES.map((p) => extensionProp(p)).join('\n')}
              </rdf:Seq>
            </pdfaSchema:property>
          </rdf:li>
        </rdf:Bag>
      </pdfaExtension:schemas>
    </rdf:Description>`;
}

interface CvExtProp {
  name: string;
  valueType: string;
  description: string;
}

const CV_EXTENSION_PROPERTIES: CvExtProp[] = [
  { name: 'version', valueType: 'Text', description: 'cvfile.org format version (MAJOR.MINOR)' },
  { name: 'created', valueType: 'Date', description: 'When the .cv file was created' },
  { name: 'modified', valueType: 'Date', description: 'When the .cv file was last modified' },
  { name: 'primaryLanguage', valueType: 'Text', description: 'BCP-47 tag of the canonical content language' },
  { name: 'primaryPayload', valueType: 'Text', description: 'Filename of the canonical text payload' },
  { name: 'generator', valueType: 'Text', description: 'Identifier of the producer' },
  { name: 'alternates', valueType: 'Text', description: 'Alternate payload descriptors (JSON-encoded array)' },
  { name: 'integrity', valueType: 'Text', description: 'Per-payload digest entries (JSON-encoded array)' },
  { name: 'embeddings', valueType: 'Text', description: 'Embedding-space summaries (JSON-encoded array)' },
];

function extensionProp(p: CvExtProp): string {
  return `                <rdf:li rdf:parseType="Resource">
                  <pdfaProperty:name>${p.name}</pdfaProperty:name>
                  <pdfaProperty:valueType>${p.valueType}</pdfaProperty:valueType>
                  <pdfaProperty:category>external</pdfaProperty:category>
                  <pdfaProperty:description>${p.description}</pdfaProperty:description>
                </rdf:li>`;
}

export function parseXmp(xml: string): CvMetadata | null {
  const cvBlock = xml;
  const get = (tag: string): string | undefined => {
    const match = cvBlock.match(new RegExp(`<cv:${tag}>([\\s\\S]*?)</cv:${tag}>`));
    return match ? xmlUnescape(match[1]!.trim()) : undefined;
  };

  const version = get('version');
  const primaryLanguage = get('primaryLanguage');
  const primaryPayload = get('primaryPayload');

  if (!version || !primaryLanguage || !primaryPayload) {
    return null;
  }

  const createdStr = get('created');
  const modifiedStr = get('modified');
  const generator = get('generator');

  const alternates = parseJsonText<AlternateMeta[]>(cvBlock, 'alternates') ?? [];
  const integrity = parseJsonText<IntegrityEntry[]>(cvBlock, 'integrity') ?? [];
  const embeddings = parseJsonText<EmbeddingSpaceSummary[]>(cvBlock, 'embeddings') ?? [];

  const meta: CvMetadata = {
    version,
    primaryLanguage,
    primaryPayload,
    alternates,
    integrity,
    embeddings,
  };
  if (createdStr) meta.created = new Date(createdStr);
  if (modifiedStr) meta.modified = new Date(modifiedStr);
  if (generator) meta.generator = generator;
  return meta;
}

function parseJsonText<T>(xml: string, tag: string): T | undefined {
  const m = xml.match(new RegExp(`<cv:${tag}>([\\s\\S]*?)</cv:${tag}>`));
  if (!m) return undefined;
  try {
    return JSON.parse(xmlUnescape(m[1]!.trim())) as T;
  } catch {
    return undefined;
  }
}

function xmlEscape(s: string): string {
  return s
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}

function xmlUnescape(s: string): string {
  return s
    .replace(/&apos;/g, "'")
    .replace(/&quot;/g, '"')
    .replace(/&gt;/g, '>')
    .replace(/&lt;/g, '<')
    .replace(/&amp;/g, '&');
}

function isoDate(d: Date): string {
  return d.toISOString().replace(/\.\d{3}Z$/, 'Z');
}
